cmake/package_util.cmake

```diff
@@ -0,0 +1,164 @@
+    PARSE_ARGV
+    0
+    arg
+    "${options}"
+    "${oneValueArgs}"
+    "${multiValueArgs}"
+  )
+
+  # make sure this is always unset at the start of the function
+  unset(cJSON_SOURCE_DIR)
+
+  project(
+    ${arg_PACKAGE_NAME}
+    VERSION ${arg_PACKAGE_VERSION}
+    DESCRIPTION ${arg_PACKAGE_DESCRIPTION}
+    HOMEPAGE_URL https://github.com/atsign-foundation/at_c
+    LANGUAGES C
+  )
+
+  if(NOT ESP_PLATFORM)
+    include(GNUInstallDirs)
+
+    # Determine if project is being built as a subproject using add_subdirectory()
+    if(arg_PACKAGE_DIR STREQUAL CMAKE_SOURCE_DIR)
+      set(PACKAGE_AS_SUBPROJECT OFF)
+    else()
+      set(PACKAGE_AS_SUBPROJECT ON)
+    endif()
+
+    message(
+      STATUS
+      "[${arg_PACKAGE_NAME}] PACKAGE_AS_SUBPROJECT: ${PACKAGE_AS_SUBPROJECT}"
+    )
+
+    foreach(package ${arg_DEPS})
+      set(${package}_DIR ${ATSDK_CMAKE_DIR}/../packages/${package})
+      include(${ATSDK_CMAKE_DIR}/${package}.cmake)
+    endforeach()
+
+    foreach(package ${arg_EXTERNAL_DEPS})
+      include(${ATSDK_CMAKE_DIR}/find_${package}.cmake)
+    endforeach()
+
+    # Create library targets
+    add_library(${PROJECT_NAME} STATIC ${arg_PACKAGE_SOURCES})
+
+    # LINK
+    # Link include headers to library targets
+    target_include_directories(
+      ${PROJECT_NAME}
+      PUBLIC
+        $<BUILD_INTERFACE:${arg_INCLUDE_DIR}>
+        $<BUILD_INTERFACE:${cJSON_SOURCE_DIR}>
+        $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
+    )
+
+    if(CMAKE_BUILD_TYPE STREQUAL "Debug")
+      target_include_directories(
+        ${PROJECT_NAME}
+        PUBLIC $<BUILD_INTERFACE:${arg_PACKAGE_DIR}/src>
+      )
+    endif()
+
+    # Link dependencies to library targets
+    target_link_libraries(${PROJECT_NAME} PUBLIC ${arg_INSTALL_TARGETS})
+
+    # INSTALL
+    # Install the include headers
+    install(
+      DIRECTORY ${arg_INCLUDE_DIR}/${PROJECT_NAME}
+      DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+    )
+
+    # Install libraries to config target
+    install(
+      TARGETS ${PROJECT_NAME} ${arg_INSTALL_TARGETS}
+      EXPORT ${PROJECT_NAME}-config
+      ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+    )
+
+    # EXPORT
+    if(NOT PACKAGE_AS_SUBPROJECT)
+      # Export the library
+      export(PACKAGE ${PROJECT_NAME})
+      # install as a config.cmake
+      install(
+        EXPORT ${PROJECT_NAME}-config
+        NAMESPACE ${PROJECT_NAME}::
+        DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME}
+        FILE ${PROJECT_NAME}-config.cmake
+      )
+
+      # export the config.cmake
+      export(
+        EXPORT ${PROJECT_NAME}-config
+        NAMESPACE ${PROJECT_NAME}::
+        FILE "cmake/${PROJECT_NAME}-config.cmake"
+      )
+    endif()
+
+    # Build the tests
+    if(arg_BUILD_TESTS)
+      enable_testing()
+      add_subdirectory(${arg_PACKAGE_DIR}/tests)
+    endif()
+  endif()
+endfunction()
+
+# add an espidf component to the build
+function(add_atsdk_espidf_component)
```

Comment:
Function for building package as an ESPIDF component (not tested, there may be some issues)